From b037e36f30a73ccb8435648d616ad6498d781fc6 Mon Sep 17 00:00:00 2001
From: Mikulas Patocka <mpatocka@redhat.com>
Date: Fri, 30 May 2014 14:54:59 -0400
Subject: [fs] dcache: prepend_path() needs to reinitialize dentry/vfsmount/mnt on restarts

Message-id: <alpine.LRH.2.02.1405301053250.29254@file01.intranet.prod.int.rdu2.redhat.com>
Patchwork-id: 81247
O-Subject: [PATCH 1/3 RHEL7 bz1092746] prepend_path() needs to reinitialize dentry/vfsmount/mnt on restarts
Bugzilla: 1092746
RH-Acked-by: Sachin Prabhu <sprabhu@redhat.com>
RH-Acked-by: Steven Whitehouse <swhiteho@redhat.com>
RH-Acked-by: David Howells <dhowells@redhat.com>
RH-Acked-by: Alexander Viro <aviro@redhat.com>

commit ede4cebce16f5643c61aedd6d88d9070a1d23a68
Author: Al Viro <viro@zeniv.linux.org.uk>
Date:   Wed Nov 13 07:45:40 2013 -0500

    prepend_path() needs to reinitialize dentry/vfsmount/mnt on restarts

    ... and equivalent is needed in 3.12; it's broken there as well

    Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>

error = 0 came from another upstream patch, but it is needed there for
correctness.

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/dcache.c b/fs/dcache.c
index d5dc456..516a49a 100644
--- a/fs/dcache.c
+++ b/fs/dcache.c
@@ -2718,9 +2718,9 @@ static int prepend_path(const struct path *path,
    const struct path *root,
    char **buffer, int *buflen)
 {
- struct dentry *dentry = path->dentry;
- struct vfsmount *vfsmnt = path->mnt;
- struct mount *mnt = real_mount(vfsmnt);
+ struct dentry *dentry;
+ struct vfsmount *vfsmnt;
+ struct mount *mnt;
  int error = 0;
  unsigned seq = 0;
  char *bptr;
@@ -2730,6 +2730,10 @@ static int prepend_path(const struct path *path,
 restart:
  bptr = *buffer;
  blen = *buflen;
+ error = 0;
+ dentry = path->dentry;
+ vfsmnt = path->mnt;
+ mnt = real_mount(vfsmnt);
  read_seqbegin_or_lock(&rename_lock, &seq);
  while (dentry != root->dentry || vfsmnt != root->mnt) {
   struct dentry * parent;
-- 
1.7.1